### PR TITLE
Fix Page Animations

### LINKED ISSUE #2335: Modal animation is inconsistent
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
  The following things need to be fixed:
  1. Search page Animations.
  2. New Group page Animations.
  3. New Chat Page Animations.
  4. Settings Page Animations.
  5. IOU Modal Page Animations.
  
## Expected Result:
 Animations for all the models & EMOJI Picker should be smooth.

## Actual Result:
 When we close RightSiderbarModal (All the Pages mentioned above), it quickly closes but the opening animation is very smooth. Sometimes, Modals are open suddenly without animation.

## Action Performed:
1. Click search which opens the search modal and notices the Opening animation.
2. Now close the modal, Modal closes quickly.

## Workaround:
Does not affect functionality.

## Platform:

 Web - :heavy_check_mark: 
 iOS - :heavy_check_mark: 
 Android - :heavy_check_mark: 
 Desktop App - :heavy_check_mark: 
 Mobile Web - :heavy_check_mark: 

**Version Number:** Latest
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/24370807/114252237-d10d9c00-99c1-11eb-872f-adfe42a34751.mp4

#### Other Video

https://expensify.slack.com/files/U01J5Q81BRP/F01SZK4SMGT/5e8bd204-81ec-4646-8d02-948b7b8048e3.webm

**Expensify/Expensify Issue URL:**

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review.

### Details
Details can be read https://github.com/Expensify/Expensify.cash/issues/2335#issuecomment-820618607

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #2335 

### Tests / QA Steps
1. Click the search icon to open the Search Modal.
2. Open all the available models from the FAB button.
3. Open the User Details / Group participants modal.
4. Open the User Settings Modal.
5. **For all the modals enter and exit animation should be smooth and consistent.**

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
   

https://user-images.githubusercontent.com/24370807/117487785-29b46280-af89-11eb-9cd5-48aa160493e8.mp4

 

#### Mobile Web

https://user-images.githubusercontent.com/24370807/116024431-cf073680-a66b-11eb-93c0-a452dbd8fdff.mp4

#### Desktop

https://user-images.githubusercontent.com/24370807/116080015-2af6ad00-a6b6-11eb-9960-4e2fef0b2232.mp4

#### iOS

 Not able to record it as I have MAC VM but the animation on IOS works smoothly.
#### Android

https://user-images.githubusercontent.com/24370807/117493077-1c4ea680-af90-11eb-94eb-4000638bf4c6.mp4

### CODE REVIEW

**Comment by parasharrajat** [comment]
Oh, Yeah. I checked somewhere the issue is linked with StackNavigator. but let me try to completely mimic the code first and grab more intel on it. :detective:

**Comment by parasharrajat** [comment]
@marcaaron I have not found any strong reason why its not working on IOS. But it seems this could be it https://github.com/software-mansion/react-native-screens/issues/713. 

I have another solution here. We can use `InteractionManager` for the Native devices. I have gone through the codebase and it works as expected with `transitionEnd` for our case. 

In short, `InteractionManager` for native devices and `transitionEnd` for Web,desktop,m-web.

Please let me know, what should be the best away forward here. Thanks.

**Comment by marcaaron** [comment]
Ok. That seems fine to me, but it would be nice if we can avoid creating two separate `ScreenWrapper` components - maybe we can create something like this...

```
.
└── libs/
    └── onScreenTransition/
        ├── index.js
        └── index.ios.js
```

```js
this.unsubscribeScreenTransitionHandler = onScreenTransition(navigation, this.callback);

componentWillUnmount() {
    this.unsubscribeScreenTransitionHandler();
}
```

I think if we go this route we should also create a new issue to track this open issue so we can clean up the exception made for iOS.

**Comment by parasharrajat** [comment]
Yeah, I was planning to create a common event listener instead of ScreenWrapper. Thanks for directions. 
I will create the follow issue as well

**Review by ctkochan22 — APPROVED** [review]

File: src/components/ScreenWrapper.js [hunk 1]
```diff
@@ -24,9 +25,15 @@ const propTypes = {
...
     includePaddingTop: PropTypes.bool,
 
-    /** react-navigation object that will allow us to goBack() */
+    // Called when navigated Screen's transition is finished.
+    onTransitionEnd: PropTypes.func,
+
+    // react-navigation navigation object available to screen components
     navigation: PropTypes.shape({
-        /** Returns to the previous navigation state e.g. if this is inside a Modal we will dismiss it */
+        // Method to attach listner to Navigaton state.
```
Comment (marcaaron) [inline src/components/ScreenWrapper.js:33]: >**listener** to **Navigation** state.

File: src/libs/onScreenTransitionEnd/index.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * Call the callback after screen transiton has ended
+ *
+ * @param {Object} navigation Screen navigation prop
+ * @param {Function} callback Method to call
+ * @returns {Function}
+ */
+function onScreenTransitionEnd(navigation, callback) {
+    return navigation.addListener('transitionEnd', evt => Str.result(callback, evt));
```
Comment (marcaaron) [inline src/libs/onScreenTransitionEnd/index.js:11]: Why `Str.result()` vs `navigation.addListener('transitionEnd', callback)` ?

**Review by marcaaron — COMMENTED** [review]
Just leaving a couple follow up comments.

File: src/libs/onScreenTransitionEnd/index.js (hunk 2 above)
Comment (parasharrajat) [reply src/libs/onScreenTransitionEnd/index.js:11]: Just to skip the test if callback is present. But yeah we can jus pass the callback directly.

Comment (parasharrajat) [reply src/libs/onScreenTransitionEnd/index.js:11]: I will add these in a new PR.

Comment (parasharrajat) [reply src/libs/onScreenTransitionEnd/index.js:11]: dropped a PR #3228. Thanks.

**Pull request merged**